framework/master-key-provider-keyring.md

```diff
@@ -0,0 +1,125 @@
+  - The string field KeyNamespace contains the provider id of the primary master key.
+  - The string field KeyName contains the key id of the primary master key.
+  - The [flags](structures.md#flags) field of this record includes the following flags:
+    - [GENERATED DATA KEY](structures.md#flags)
+    - [ENCRYPTED DATA KEY](structures.md#flags)
+    - [SIGNED ENCRYPTION CONTEXT](structures.md#flags) (if the master key supports encryption context signing)
+
+Next, for each [master key](master-key-interface.md) provided by the master key provider's 
+[Get Master Keys for Encryption](master-key-provider-interface.md#get-master-keys-for-encryption), the keyring MUST call
+ [EncryptDataKey](master-key-interface.md#encrypt-data-key). If the primary master key was used to generate a data key,
+ that master key MUST NOT be used again for encryption.
+ 
+ The input to [EncryptDataKey](master-key-interface.md#encrypt-data-key) MUST contain the following from the [encryption materials](structures.md#encryption-materials):
+ - [Algorithm Suite](structures.md#algorithm-suite)
+ - [Encryption Context](structures.md#encryption-context)
+ - [Plaintext Data Key](structures.md#plaintext-data-key)
+ 
+If any master key's [EncryptDataKey](master-key-interface.md#encrypt-data-key) fails, OnEncrypt MUST fail.
+
+For each master key, OnEncrypt MUST do the following with the response from [Encrypt Data Key](master-key-interface.md#encrypt-data-key)
+- append a new [encrypted data key](structures.md#encrypted-data-key) to the encrypted data key list
+  in the [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - the [ciphertext](structures.md#ciphertext) is the response encrypted data key.
+  - the [key provider id](structures.md#key-provider-id) is the provider id of the master key.
+  - the [key provider information](structures.md#key-provider-information) is the key id of the master key.
+- append a new record to the [keyring trace](structures.md#keyring-trace)
+  in the input [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - The string field KeyNamespace contains the provider id of the master key.
+  - The string field KeyName contains the key id of the master key.
+  - The [flags](structures.md#flags) field of this record includes the following flags:
+    - [ENCRYPTED DATA KEY](structures.md#flags)
+    - [SIGNED ENCRYPTION CONTEXT](structures.md#flags) (if the master key supports encryption context signing)
+
+### OnDecrypt
+
+If the input [decryption materials](structures.md#decryption-materials) contain a plaintext data key,
+OnDecrypt MUST immediately return the unmodified decryption materials.
+
+Otherwise, OnDecrypt MUST attempt to decrypt the [encrypted data keys](structures.md#encrypted-data-key)
+in the input using [Decrypt Data Key](master-key-provider-interface.md#decrypt-data-key)
```

Comment:
They also MUST be attempted in order.

Reply:
I've specified now that the input to DecryptDataKey must contain the Encrypted Data Keys in their original order